File: Java/Array/3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        HashSet<List<Integer>> res = new HashSet<>();
        Arrays.sort(nums);

        if (nums.length < 3) return new ArrayList<>(res);
        
        for (int i = 0; i < nums.length - 2; i++) {
            int curr = nums[i];

            if (curr > 0) break;

            int slow = i + 1;
            int fast = nums.length - 1;
            while (slow < fast) {
                int addition = curr + nums[slow] + nums[fast];
                if (addition < 0) {
                    slow++;
                } else if (addition > 0) {
                    fast--;
                } else {
                    res.add(Arrays.asList(nums[i],nums[slow],nums[fast]));
                    slow++;
                    fast--;
                }
            }
        }

        return new ArrayList<>(res);
    }
}

File: Python/Array/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        indexMap = {}
        nums.sort() # nlogn

        for i in range(0, len(nums)):
            num = nums[i]
            indexMap[num] = i

        res = []
        done = set()
        for i in range(len(nums)):
            num1 = nums[i]
            for j in range(i+1, len(nums)): # n^2
                num2 = nums[j]
                if ((num1, num2) in done):
                    continue
                num3 = 0 - num1 - num2
                k = indexMap.get(num3, None)
                if k and k > j:
                    done.add((num1, num2))
                    res.append([num1, num2, num3])
                    continue
        
        return list(res)
                    